
<!-- <img src="man/figures/logo.png" align="right" alt="" width="120" /> -->
<!-- README.md is generated from README.Rmd. Please edit that file -->

# A2TEA.WebApp <img src="man/figures/logo.png" align="right" alt="" width="120" />

<!-- badges: start -->

[![R-CMD-check](https://github.com/tgstoecker/A2TEA.WebApp/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/tgstoecker/A2TEA.WebApp/actions/workflows/R-CMD-check.yaml)
[![](https://img.shields.io/github/last-commit/tgstoecker/A2TEA.WebApp.svg)](https://github.com/tgstoecker/A2TEA.WebApp/commits/master)
[![Lifecycle:
stable](https://img.shields.io/badge/lifecycle-stable-brightgreen.svg)](https://www.tidyverse.org/lifecycle/#stable)
[![DOI](https://zenodo.org/badge/525064928.svg)](https://zenodo.org/badge/latestdoi/525064928)

<!-- badges: end -->

## **02.11.2023: The A2TEA.WebApp is is no longer maintained in this repository - all future development will happen in [https://github.com/groupschoof/A2TEA.WebApp](https://github.com/groupschoof/A2TEA.WebApp)**

A2TEA is a software workflow facilitating identification of candidate
genes for stress adaptation based on comparative genomics and
transcriptomics. It combines differential gene expression with gene
family expansion as an indicator for the evolution of adaptive traits.  
The goal of the A2TEA.WebApp is to allow exploration, highlighting, and
exporting of the results generated by a A2TEA pipeline run. The core of
the package is a Shiny application that aims to combine phylogenetic,
transcriptomic & functional data and results, in a way that makes it
easier to generate insightful observations. As the A2TEA pipeline
requires the user to formulate hypotheses - for each of which specific
results are produced - the WebApp allows to investigate the genomic
adaptations of one or a subset of the investigated species to the
analyzed treatment condition. The A2TEA.WebApp aims to combine the
benefits of interactivity and reproducibility, e.g. by capturing genes
or orthologous groups of interest bookmarked by the user and allowing
for easy export of plots, tables and complete subsets of the original
input data. The newest releases of the workflow and WebApp now allow for
final results without the need for transcriptomic data for all species.
This way, valuable phylogenetic information from species without
available expression information for the conditions under investigation
can be included.

**See the published paper!:**
<https://f1000research.com/articles/11-1137>

<br>

## :arrow_double_down: Download & Setup

There are currently 3 ways of using the A2TEA.WebApp:  
- installing the development version as an R package from this
repository with devtools/remotes  
- a docker container of the latest stable release, for which we offer a
singularity based guide  
- a demo instance of the App running hosted on
[shinyapps.io](https://www.shinyapps.io/) allowing for a sneak peak of
the functionality

### Option 1: Installation

You can install the development version of A2TEA.WebApp from GitHub with
either remotes or devtools, e.g.:

``` r
library(remotes)
remotes::install_github("tgstoecker/A2TEA.WebApp", 
                        dependencies = TRUE, 
                        build_vignettes = TRUE)
```

Note that some system dependencies might have to be installed:

``` bash
apt-get update
apt-get upgrade
apt-get install -y \
    libxml2-dev \
    libcairo2-dev \
    libgit2-dev \
    default-libmysqlclient-dev \
    libpq-dev \
    libsasl2-dev \
    libsqlite3-dev \
    libssh2-1-dev \
    libxtst6 \
    libcurl4-openssl-dev \
    unixodbc-dev \
    libproj-dev xdg-utils \
    --fix-missing
```

You might also need to install pandoc:  
- either in a terminal with conda/mamba
`conda/mamba install -c conda-forge pandoc`  
- or via [installer/package manager](https://pandoc.org/installing.html)

### Option 2 - Docker container using singularity

You can also circumvent dependency issues by downloading the latest
release of our [A2TEA Docker
image](https://hub.docker.com/repository/docker/tgstoecker/a2tea_webapp).  
To circumvent most potential problems regarding access rights we use
Singularity to pull and use the image.  
Note however, that we require `--writable` to be enabled and therefore
most likely require either the `--fakeroot` or `--no-home` flag as well.
Which command works will depend on your Singularity installation -
whether or not it was priviliged or not.  
If you have sudo rights and installed Singularity via your systems
package manager then there should be no problem at all.

``` bash
#pull the image from dockerhub
singularity pull a2tea_webapp.sif docker://tgstoecker/a2tea_webapp:latest

#open R console of image in non-persistent but writable mode
singularity run --writable --fakeroot a2tea_webapp.sif R
#or
singularity run --writable --no-home a2tea_webapp.sif R
```

#### Starting the App

For both option 1 and 2 starting the Shiny Application requires loading
the library and calling the A2TEA_App function:

``` r
library(A2TEA.WebApp)
A2TEA_App()
#upload an A2TEA.RData file
```

### Option 3 - Hosted version on shinyapps.io

Click on this link - <https://tgstoecker.shinyapps.io/A2TEA-WebApp/> -
and a new browser tab will open in which a running instance of the
WebApp will be started.

<br>

## Usage overview

All outlined setup options of the App come with an included demo dataset
which can be loaded via a single button press. You can find the rendered
version of the documentation of `A2TEA.WebApp` at the project website
<https://tgstoecker.github.io/A2TEA.WebApp>, created with `pkgdown`,
which includes a usage guide detailing all of the App’s functionality.

<br>

## :ballot_box_with_check: Included test data

The WebApp comes with a test dataset and can be loaded via clicking the
“Try a demo A2TEA.RData file” at the top of the interface. This result
is also computed by the
[A2TEA.Workflow](https://github.com/tgstoecker/A2TEA.Workflow) for its
usage test. The workflow is set up to run a three species analysis with
*Hordeum vulgare*, *Zea mays* & *Oryza sativa japonica* and their
reaction patterns to drought stress.  
Peptide fastas are reduced to 2000 proteins; sequencing reads are
subsampled to 2M reads.  
The fasta/annotation files and sequencing reads are hosted - here.  
Fasta & annotation files are all either downloaded from ensemblPlants
and still possess their original name or in the case of the functional
annotations were computed using
[AHRD](https://github.com/groupschoof/AHRD).  
NCBI SRA accession IDs of sequencing reads: - Hordeum vulgare:
SRR6782243, SRR6782247, SRR6782257, SRR6782249, SRR6782250, SRR6782254 -
Zea mays: SRR2043219, SRR2043217, SRR2043190, SRR2043220, SRR2043226,
SRR2043227 - Oryza sativa japonica: SRR5134063, SRR5134064, SRR5134065,
SRR5134066

These correspond to the following studies relating to drought stress:  
- Hordeum vulgare: <https://doi.org/10.1186/s12864-019-5634-0>  
- Zea mays: <https://doi.org/10.1104/pp.16.01045>  
- Oryza sativa japonica: <https://doi.org/10.3389/fpls.2017.00580>

## Development

If you encounter a bug, have a usage questions, or want to share ideas
and functionality to make this package better, feel free to file an
[issue](https://github.com/tgstoecker/A2TEA.WebApp/issues).

<br>

## License

MIT © Tyll Stöcker
